utilities/object_registry.cc

```diff
@@ -26,61 +55,412 @@ const ObjectLibrary::Entry *ObjectLibrary::FindEntry(
+  }
+
+  // Makes a copy of this registry.
+  std::shared_ptr<ObjectRegistry> Clone() const override {
+    return std::make_shared<ObjectRegistryImpl>(*this);
+  }
+
+  // Returns the registry contents as a printable string, suitable for Dump
+  std::string GetPrintableOptions() const override {
+    size_t types, factories;
+    factories = GetFactoryCount(&types);
+    const int kBufferSize = 200;
+    char buffer[kBufferSize];
+    snprintf(buffer, kBufferSize, "%" ROCKSDB_PRIszt "/%" ROCKSDB_PRIszt "\n",
+             types, factories);
+    std::string result = buffer;
+
+    for (const auto &library : options_.libraries) {
+      result.append(library->GetPrintableOptions() + "\n");
+    }
+    return result;
+  }
+
+  void AddLocalLibrary(const RegistrarFunc &registrar,
+                       const std::string &method,
+                       const std::string &arg) override;
+
+  // Creates a new local library, registering the factories in method
+  Status AddLocalLibrary(Env *env, const std::string &method,
+                         const std::string &arg) override;
+
+  // Loads the types from the named library and method into a new dynamic object
+  Status AddDynamicLibrary(Env *env, const std::string &library,
+                           const std::string &method,
+                           const std::string &arg) override;
+
+  // Loads the method from the named library and method into a new dynamic
+  // object library.
+  Status AddDynamicLibrary(const std::shared_ptr<DynamicLibrary> &library,
+                           const std::string &method,
+                           const std::string &arg) override;
+
+  // Returns the number of registered types for this registry.
+  // If specified (not-null), types is updated to include the names of the
+  // registered types.
+  size_t GetRegisteredTypes(
+      std::unordered_set<std::string> *types) const override {
+    assert(types);
+    for (const auto &library : options_.libraries) {
+      library->GetRegisteredTypes(types);
+    }
+    return types->size();
+  }
+
+  // Returns the number of registered names for the input type
+  // If specified (not-null), names is updated to include the names for the type
+  size_t GetRegisteredNames(const std::string &type) const override {
+    return GetRegisteredNames(type, nullptr);
+  }
+
+  size_t GetRegisteredNames(const std::string &type,
+                            std::vector<std::string> *names) const override {
+    size_t count = 0;
+    for (const auto &library : options_.libraries) {
+      count += library->GetRegisteredNames(type, names);
+    }
+    return count;
+  }
+
+  // Returns the total number of factories registered for this library.
+  // This method returns the sum of all factories registered for all types.
+  // @param num_types returns how many unique types are registered.
+  size_t GetFactoryCount(size_t *num_types) const override {
```

Comment:
Make it clear that, the returned value is the sum of all factories registered for all types, "num_types" only contains the types